### PR TITLE
Fix: Include additional validations to compliance

### PR DESCRIPTION
**Pull request by joaogularte**
The main goal is include additional validations according BACEN tester software uses cases following the specification [EMV® QR Code Specification for Payment Systems (EMV® QRCPS) Merchant-Presented Mode](https://www.emvco.com/emv-technologies/qrcodes/). 

Validations can be summarize as follows:
* QR Code size respecting 512 bytes maximum
* Alphanumeric Special format for fields that follow it
* Remove zero acceptance for transaction amount
* UUID format for random key
* Ensure no protocol presence on URL

### CODE REVIEW

File: lib/ex_pix_brcode/brcode/decoder.ex [hunk 1]
```diff
@@ -90,14 +94,14 @@ defmodule ExPixBRCode.BRCodes.Decoder do
...
-  defp extract_crc(binary) when byte_size(binary) > 4 do
+  def extract_crc(binary) when byte_size(binary) > 4 do
     len = byte_size(binary)
     crc = binary_part(binary, len, -4)
     val = binary_part(binary, 0, len - 4)
     {:ok, {val, crc}}
   end
 
-  defp extract_crc(_binary), do: {:error, :invalid_input_length}
+  def extract_crc(_binary), do: {:error, :invalid_input_length}
```
Comment (polvalente) [inline lib/ex_pix_brcode/brcode/decoder.ex:104]: Is this change really needed?

File: lib/ex_pix_brcode/brcode/models/brcode.ex [hunk 2]
```diff
@@ -162,12 +182,12 @@ defmodule ExPixBRCode.BRCodes.Models.BRCode do
   end
 
   defp validate_url(changeset, url) do
-    case URI.parse("https://" <> url) do
-      %{path: path} when is_binary(path) ->
+    with {:validate_has_web_protocol, nil} <- {:validate_has_web_protocol, Regex.run(~r/^[https:\/\/]+/, url)},
```
Comment (polvalente) [inline lib/ex_pix_brcode/brcode/models/brcode.ex:185]:
```suggestion
    with {:validate_has_web_protocol, nil} <- {:validate_has_web_protocol, Regex.run(~r{^[https://]+}, url)},
```

Comment (polvalente) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]: Suggestion for using `{}` as delimiters, but also because "hhh" passes the provided regex

File: lib/ex_pix_brcode/payments/models/static_pix_payment.ex [hunk 3]
```diff
@@ -28,5 +28,15 @@ defmodule ExPixBRCode.Payments.Models.StaticPixPayment do
     |> validate_length(:transaction_amount, max: 13)
     |> validate_format(:transaction_amount, ~r/(^0$)|(^[0-9]+\.[0-9]*$)/)
     |> validate_format(:transaction_id, ~r/^[a-zA-Z0-9]{1,25}$/)
+    |> validate_random_key_format()
+  end
+
+  defp validate_random_key_format(changeset) do
+    key_type = get_change(changeset, :key_type)
+    if key_type == "random_key" do
+      validate_format(changeset, :key, ~r/[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}/)
```
Comment (polvalente) [inline lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: Instead of using a Regex, I think you should use either `:ecto` or `:uuid` to validate that it actually is a valid UUID4

File: test/ex_pix_brcode/decoder_test.exs [hunk 4]
```diff
@@ -361,28 +361,53 @@ defmodule ExPixBRCode.DecoderTest do
...
+      assert {:error, {:validation, changeset}} = Decoder.decode_to("00020126830014br.gov.bcb.pix01364004901d-bd85-4769-8e52-cb4c42c506dc0221Jornada pagador 57768520400005303986540573.625802BR5903Pix6008Brasília62080504oooo6304B243")
+      assert [merchant_city: {"has invalid format", [validation: :format]}] == changeset.errors
+    end
+
+    test "succeeds on validate invalid format for reference_label BRCode field" do
+      assert {:error, {:validation, changeset}} = Decoder.decode_to("00020104141234567890123426580014BR.GOV.BCB.PIX0136123e4567-e12b-12d1-a456-42665544000027300012BR.COM.OUTRO011001234567895204000053039865406123.455802BR5917NOME DO RECEBEDOR6008BRASILIA610870074900622005161234567890👀1234580390012BR.COM.OUTRO01190123.ABCD.3456.WXYZ630475EF")
+      assert [reference_label: {"has invalid format", [validation: :format]}] = changeset.changes.additional_data_field_template.errors
+    end
+
+    test "Ssucceeds on validate invalid format for reference_label BRCode field" do
```
Comment (polvalente) [inline test/ex_pix_brcode/decoder_test.exs:408]:
```suggestion
    test "Succeeds on validate invalid format for reference_label BRCode field" do
```

File: lib/ex_pix_brcode/brcode/decoder.ex (hunk 1 above)
Comment (renanlage) [reply lib/ex_pix_brcode/brcode/decoder.ex:104]: If it is, there should be a typespec for it

Comment (polvalente) [reply lib/ex_pix_brcode/brcode/decoder.ex:104]: And if it's only for tests, `@doc false` and no typespec is the way to go

Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/decoder.ex:104]: > Is this change really needed?

Actually not. I've changed it to help me during development process. I will back to `defp`.

File: lib/ex_pix_brcode/brcode/models/brcode.ex (hunk 2 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]: I totally agree with suggestion about use `match?` instead of `run`. However I am not sure about the use of `{}` as delimiters. What did you mean with ""hhh" passes the provided regex"?

Comment (polvalente) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]:
```elixir
iex(1)> Regex.match?(~r/^[https:\/\/]+/, "hhh")
true
iex(2)> Regex.match?(~r/^[https:\/\/]+/, "hhhsss")
true
iex(3)> Regex.match?(~r/^[https:\/\/]+/, "hhhsss///")
true
iex(4)> Regex.match?(~r{^https://\w+}, "hhhsss///")  
false
iex(5)> Regex.match?(~r{^https://\w+}, "hhh")      
false
iex(6)> Regex.match?(~r{^https://\w+}, "http")
false
iex(7)> Regex.match?(~r{^https://\w+}, "https")
false
iex(8)> Regex.match?(~r{^https://\w+}, "https://")
false
iex(9)> Regex.match?(~r{^https://\w+}, "https://s")
true
```

Comment (polvalente) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]: `{}` as delimiters add a lot of legibility to the regex IMO

The problem with your original regex is that you used a character class instead of a capturing `(...)` or a non-capturing `(?:..)` group. However, none of those are needed because the string itself doesn't repeat and isn't optional or anything.

Btw, `http://` isn't accepted. Is this intentional? If not, add a `?` right after `s` in my suggestion

Comment (renanlage) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]: > However I am not sure about the use of {}

I think it avoids the need to escape "/" like you did

> What did you mean with ""hhh" passes the provided regex"?

You used `[]` enclosing the "https://" regex on the match, which means any character inside the brackets will match. You should remove the brackets and to a direct match.

File: lib/ex_pix_brcode/brcode/decoder.ex [hunk 5]
```diff
@@ -74,12 +76,14 @@ defmodule ExPixBRCode.BRCodes.Decoder do
               | {:unknown_key, String.t()}}
              | :unknown_error
              | :invalid_crc
-             | :invalid_input_length}
+             | :invalid_input_length
```
Comment (renanlage) [inline lib/ex_pix_brcode/brcode/decoder.ex:79]: In what situation this error is returned?

File: lib/ex_pix_brcode/brcode/models/brcode.ex [hunk 6]
```diff
@@ -78,9 +83,16 @@ defmodule ExPixBRCode.BRCodes.Models.BRCode do
     |> validate_inclusion(:transaction_currency, ~w(986))
     |> validate_length(:transaction_amount, max: 13)
     # Formats accept: "0", "0.10", ".10", "1.", "1", "123.9","123.99", "123456789.23"
-    |> validate_format(:transaction_amount, ~r/^0$|^[0-9]+\.[0-9]{2}$|^[0-9]+\.[0-9]{1}$|^[1-9]{1}[0-9]*\.?$|^\.[0-9]{2}$/)
+    |> validate_format(
+      :transaction_amount,
+      ~r/^[0-9]+\.[0-9]{2}$|^[0-9]+\.[0-9]{1}$|^[1-9]{1}[0-9]*\.?$|^\.[0-9]{2}$/
```
Comment (renanlage) [inline lib/ex_pix_brcode/brcode/models/brcode.ex:88]: I see that you removed "0" from the regex. Is it because it is no longer supported ou because it is supported by one of the other regexes?

File: lib/ex_pix_brcode/brcode/models/brcode.ex [hunk 7]
```diff
@@ -117,6 +129,14 @@ defmodule ExPixBRCode.BRCodes.Models.BRCode do
     |> cast(params, [:reference_label])
     |> validate_required([:reference_label])
     |> validate_length(:reference_label, min: 1, max: 25)
```
Comment (renanlage) [inline lib/ex_pix_brcode/brcode/models/brcode.ex:131]: Is this the txid? Isn't the max size 35 instead of 25?

File: lib/ex_pix_brcode/payments/models/static_pix_payment.ex (hunk 3 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: Yep, this would be a more elegant solution, but I didn't use this kind because of these behaviors:  

This uuid `4004901D-BD85-4769-8E52-CB4C42C506DC`, that Bacen considers invalid, is cast by `Ecto.UUID.cast/1` as followed: 

```elixir
Ecto.UUID.cast("4004901D-BD85-4769-8E52-CB4C42C506DC")
{:ok, "4004901d-bd85-4769-8e52-cb4c42c506dc"}
```

And with `UUID.info/1`:

```elixir
{:ok,
 [
   uuid: "4004901D-BD85-4769-8E52-CB4C42C506DC",
   binary: <<64, 4, 144, 29, 189, 133, 71, 105, 142, 82, 203, 76, 66, 197, 6,
     220>>,
   type: :default,
   version: 4,
   variant: :rfc4122
 ]}
```

Both cases the Bacen invalid uuid is casted or considered valid.

Comment (polvalente) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: If the problem is that only lowercase is accepted, I'd use UUID.info and a check that string == String.downcase(string)

Comment (polvalente) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: Also, "b54895cc-3b33-fca3-84b3-61cd6e2711db" is an invalid UUID4 and passes both your regex and `UUID.info`, which yields `version: 15`. If only uuid4 is accepted, you must also validate that the returned `:version` is `4`

Comment (polvalente) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: https://en.wikipedia.org/wiki/Universally_unique_identifier#:~:text=IEC%209834%2D8.-,Format,e89b%2D12d3%2Da456%2D426614174000

File: lib/ex_pix_brcode/brcode/decoder.ex (hunk 5 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/decoder.ex:79]: This error is returned when the byte size of argument passed to `defp extract_crc/1` is less than four. IHMO we can follow just with `:invalid_input_size` error, because of what is really validate by the guard is the string size and not the length.

File: lib/ex_pix_brcode/brcode/models/brcode.ex (hunk 6 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:88]: It is because it is no longer supported. Zero value isn't accept on transaction_amount field, according BACEN tests.

File: lib/ex_pix_brcode/brcode/models/brcode.ex (hunk 7 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:131]: Yes, this is the txId. The max size 35 should be for dynamic qrcode, by the way the txid for this type of qrcode should not be encoded in qrcode and if it happens should not be considered. With static qrcode the txid is required  and the max length is 25. 

Thinking better now, the cast process of `reference_label` field should consider the qrcode type. When qrcode is dynamic the `reference_label` is not cast and therefore its format is not validated. When qrcode is static the `reference_label` is cast and its format validated.

File: lib/ex_pix_brcode/payments/models/static_pix_payment.ex (hunk 3 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:37]: Thanks for the explanation. 

The problem is the occurrence of uppercase. Your suggestion of check `string == String.downcase(string)` is enough for this situation.

File: lib/ex_pix_brcode/brcode/models/brcode.ex (hunk 2 above)
Comment (joaogularte) [reply lib/ex_pix_brcode/brcode/models/brcode.ex:185]: It's clear now, thanks for the explanations guys. I am going to change the regex.

**Review by renanlage — APPROVED** [review]

**Review by polvalente — APPROVED** [review]

File: lib/ex_pix_brcode/payments/models/static_pix_payment.ex [hunk 8]
```diff
@@ -28,5 +28,17 @@ defmodule ExPixBRCode.Payments.Models.StaticPixPayment do
...
+    |> validate_random_key_format()
+  end
+
+  defp validate_random_key_format(changeset) do
+    key_type = get_change(changeset, :key_type)
+    key = get_change(changeset, :key)
+
+    cond do
+      key_type == "random_key" and key == String.downcase(key) -> changeset
+      key_type == "random_key" and key != String.downcase(key) -> add_error(changeset, :key, "has invalid format")
```
Comment (polvalente) [inline lib/ex_pix_brcode/payments/models/static_pix_payment.ex:40]:
```suggestion
      key_type == "random_key" and key == String.downcase(key) -> changeset
      key_type == "random_key" -> add_error(changeset, :key, "has invalid format")
```

To avoid running `String.downcase` twice

Comment (polvalente) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:40]: by the way, I think this is missing the UUID validation, or was it in another function?

Comment (joaogularte) [reply lib/ex_pix_brcode/payments/models/static_pix_payment.ex:40]: @polvalente the uuid validation was on a `defp` inside `ExPixBRCode.Payments` module.
